### ISSUE TITLE
Directly calling exit() to exit the app is incredibly error-prone

### ISSUE THREAD

**Issue by nneonneo** [openframeworks/openFrameworks#3065]
# tl;dr

Relying on `std::exit()` to call `ofApp::exit()` eventually risks having `ofApp::exit()` see a partially torn-down application, which can cause fun crashes on exit. We should be calling `ofApp::exit()` before `std::exit()`, so the app has full control over the exit process.
# Details

When you press esc or Cmd-Q/Alt+F4 to quit an OpenFrameworks app, it basically calls `exit` directly. `exit` runs `atexit` handlers and static destructors, and one of those handlers eventually calls `ofApp::exit`. However, atexit may run other handlers first, since it calls them in reverse order (with the most recently-registered handler first).

This is a huge problem when threads become involved. In three of my three last projects I've been bitten by a hang or crash on exit due basically to the fact that I have zero opportunity to catch an exit-in-progress before it starts happening. The problem is basically as follows:
- App starts up and registers the `ofApp::exit` handler
- App starts a thread that uses some library
- Other library creates static variables or objects, or calls atexit()
- When the app quits, the other library is deinitialized
- Thread is still running because `ofApp` has not killed it (and has no opportunity to know when to kill it)
- `ofApp::exit` is called, at which point I would usually stop the thread
- Thread either hangs (because the library is no longer sending events), crashes (because it accesses bad library state), or appears to work (if you get "lucky" and the library isn't accessed before the thread exits).

This is really nasty! It's a very terrible practice to call `ofApp::exit()` while parts of the app have already been shutdown. In the example above, it is called too late to stop the thread from crashing, and in general it can result in situations where parts of the app have been torn down by the time it is called.

It can lead to insane, hard to fix bugs, and I think many people may see that the app is crashing on exit and not caring (because, after all, it works up to that point!). But crashing on exit wastes time (crash dumps, debugger traps, 10 second hangs waiting for threads, infinite hangs if you're unlucky, etc.), and presents a bit of a usability problem ("don't worry about the app if it pops up an error when it exits!").

The solution's really easy! Just call `ofApp::exit()` (or the equivalent Poco notifier) _before_ calling `std::exit()`. Processing does this, for example. I doubt there's ever a case where `ofApp::exit()` should rely on `std::exit()` having been called, though I'd like to know if there is such a case.
## Specific cases of affected libraries
- One library I used created a `boost::spirit` parser in a thread. `boost::spirit` creates a static object which is deleted during atexit handling. The library's thread continues running, and will crash when it attempts to use the parser. The library provides a deinit function, but presently I'm calling the deinit function from `onKeyPressed(OF_KEY_ESC)` which is a gross hack (and doesn't work with Cmd-Q/Alt+F4).
- An add-on I used (`ofxOpenNI2`) creates a thread and waits for events from a library. The thread hangs when the library deinitializes during exit, making the whole app hang on exit.
